[ARDMediathekBridge] Switch to JSON-API

Mar-Koeh opened the pull request:
The html screen scraping approach of ARDMediathekBridge did not work reliably. I could not find one show for which the item list was not empty using the html screen scraping approach.

The proposed change uses the JSON-API of the WebApp. Although there is still room for improvement (feed title, better understanding of the API, more accurate mimic of the webapp's behavior, de-pagination …), it does work with this change.

Mar-Koeh committed:
Switch ARDMediathekBridge to JSON-API

The html screen scraping approach of ARDMediathekBridge did not work reliably. I could not find one show for which the item list was not empty using the html screen scraping approach.

The proposed change uses the JSON-API of the WebApp. Although there is still room for improvement (feed title, better understanding of the API, more accurate mimic of the webapp's behavior, de-pagination …), it does work with this change.

Mar-Koeh committed:
[ARDMediathekBridge] Fixed phpcs issues

[ARDMediathekBridge] Fixed phpcs issues: spaces between minus, array_key_last() not available in supported PHP version.

f0086 commented:
Tested and worked smoothly.
It took me a moment to find the right path, so maybe adjust the title of the path option so it is clear which path is needed here (all without the https://www.ardmediathek.de/sendung/ prefix).

Besides that, its good to go.

Mar-Koeh commented:
> path option so it is clear which path is needed here (all without the https://www.ardmediathek.de/sendung/ prefix).

That was the requirement of the previous version of the bridge. With this change, the whole URL, the ID with the slug (as before) and just the ID are accepted. The JSON-API that this version of the bridge switched to just needs the ID. Thus, a string extractor has been implemented that obtains the ID. Thus, it maintains compatibility to old inputs. All of the following works:

- https://www.ardmediathek.de/sendung/45-min/Y3JpZDovL25kci5kZS8xMzkx/
- 45-min/Y3JpZDovL25kci5kZS8xMzkx (backward compatibility to the previous version)
- Y3JpZDovL25kci5kZS8xMzkx

I guess I best change the label to “Show Link or ID”.

Mar-Koeh committed:
[ARDMediathekBridge] updated parameter description

Indicate that now full URLs as well as just the ID are accepted.

f0086 commented:
That's perfect! :ok_hand:

Mar-Koeh commented:
> That's perfect! 👌

Thanks a lot, @f0086 I have pushed the change to the branch. Once the workflows are done with no problems, I hope for a merge. 😊 I made no changes to the code style and the used functions are the same, so I expect the checks to be successful.

em92 commented on bridges/ARDMediathekBridge.php:
```diff
@@ -4,32 +4,87 @@ class ARDMediathekBridge extends BridgeAbstract {
-			$item['content'] = '<img src="' . $video->find('img', 0)->src . '" />';
-			$item['timestamp'] = strtotime(mb_substr($video->find('div.Line-epbftj-1', 0)->plaintext, 0, 10));
+		$url = SELF::APIENDPOINT . $showID . '/?pageSize=' . SELF::PAGESIZE;
+		$rawJSON = getContents($url) or returnServerError('Could not load content');
```
getContents will raise exception on error, so returnServerError won't execute.

Mar-Koeh replied:
Thanks a lot for your feedback! To be honest, I got this from [InstagramBridge](https://github.com/RSS-Bridge/rss-bridge/blob/76c38332ee877986f13bd64fc34b4228ee8fccec/bridges/InstagramBridge.php#L242). It is a pattern I also observed in [GOGBridge](https://github.com/RSS-Bridge/rss-bridge/blob/76c38332ee877986f13bd64fc34b4228ee8fccec/bridges/GOGBridge.php#L12), [PatreonBridge](https://github.com/RSS-Bridge/rss-bridge/blob/76c38332ee877986f13bd64fc34b4228ee8fccec/bridges/PatreonBridge.php#L172) and many others.

Which change would you suggest?

em92 replied:
> It is a pattern I also observed in GOGBridge, PatreonBridge and many others.

Long story, how this pattern still exists in many bridges. Just opened PR https://github.com/RSS-Bridge/rss-bridge/pull/2398 to take out all usages.

> Which change would you suggest?

I suggest to remove "or returnServerError"

Mar-Koeh replied:
Okay! I removed it,

em92 commented on bridges/ARDMediathekBridge.php:
```diff
@@ -4,32 +4,87 @@ class ARDMediathekBridge extends BridgeAbstract {
+		$pathComponents = explode('/', $this->getInput('path'));
+		if (empty($pathComponents)) {
+			$pathComponents = explode('/', SELF::PARAMETERS[0]['defaultValue']);
+		}
```
`path` is required parameter. Why do we need to handle case, where it is empty?

Mar-Koeh replied:
Try for example ArteBridge with these parameters `?action=display&bridge=Arte7&context=Collection+(Français)&colfr=&format=Html`. Although the parameter is required, A zero-length string can still be called in a deep-link. I found it a good practice not to access an array at keys it wouldn't have but to safely fall back to the default value.

What would be better option?

em92 replied:
I see. 
In this case I suggest to raise exception using returnClientError.

Mar-Koeh replied:
Wonderful. I'm doing that now.

em92 commented on bridges/ARDMediathekBridge.php:
```diff
@@ -4,32 +4,87 @@ class ARDMediathekBridge extends BridgeAbstract {
 	const URI = 'https://www.ardmediathek.de';
 	const DESCRIPTION = 'Feed of any series in the ARD-Mediathek, specified by its path';
 	const MAINTAINER = 'yue-dongchen';
```
90% of bridge code has been rewritten.
If you, @Mar-Koeh, are going to maintain this bridge, you can add your github name via comma.

Bridge maintainers are pinged in issues, when error happens.

Also question for @yue-dongchen, are still maintaining this bridge?

yue-dongchen replied:
I didn't know of the API before. I'll stay there and watch out for new issues.

Mar-Koeh replied:
> 90% of bridge code has been rewritten. If you, @Mar-Koeh, are going to maintain this bridge, you can add your github name via comma.

if @yue-dongchen is fine with it, I'd join maintaining this bridge.

Mar-Koeh committed:
Removed redundant or returnServerError

according to @em92 returnServerError would never be called as the other function either works successfully or throws an exception.

Mar-Koeh committed:
Show error message instead of default feed

As suggested by @em92 an empty required parameter `path `now causes an error message
